File: c++/words/src/words.cpp
#include "../include/words.h"

#include <string>
#include <vector>
#include <sstream>
#include <algorithm>

std::string Words::reverse_string(const std::string& str) {
    int len_str = str.length();
    std::ostringstream buffer;
    for(int i = 0; i < len_str; i++) {
        buffer << str[len_str - 1 - i];
    }
    
    return buffer.str();
}

bool Words::is_palindrome(const std::string& str) {
    std::string local_str = str;
    local_str.erase(std::remove_if(local_str.begin(), local_str.end(), ::isspace), local_str.end());
    int i = 0;
    int j = local_str.length() - 1;
    while(i < j) {
        if(local_str[i++] != local_str[j--]) return false;
    }
    
    return true;
}

std::vector<int> Words::compute_lps_array(const std::string& pattern) {
    int len_pattern = pattern.length();
    std::vector<int> lps(len_pattern, 0);
    int length = 0;
    int i = 1;

    while(i < len_pattern) {
        if(pattern[i] == pattern[length]) {
            length++;
            lps[i] = length;
            i++;
        } else {
            if (length != 0) {
                length = lps[length - 1];
            } else {
                lps[i] = 0;
                i++;
            }
        }
    }

    return lps;
}

bool Words::kmp_search(const std::string& sub, const std::string& str) {
    int len_sub = sub.length();
    int len_str = str.length();
    if(len_sub == 0) return true;
    else if(len_sub > len_str) return false;

    std::vector<int> lps = compute_lps_array(sub);
    int i = 0;
    int j = 0;
    while(i < len_str) {
        if(str[i] == sub[j]) {
            i++;
            j++;
            if(j == len_sub) {
                return true;
            }
        } else {
            if(j != 0) {
                j = lps[j - 1];
            } else {
                i++;
            }
        }
    }

    return false;
}

int Words::lcs(const std::string& str1, const std::string& str2) {
    int len_str1 = str1.length();
    int len_str2 = str2.length();
    std::vector<std::vector<int>> dp(len_str1 + 1, std::vector<int>(len_str2 + 1, 0));

    for(int i = 1; i <= len_str1; i++) {
        for(int j = 1; j <= len_str2; j++){
            if(str1[i - 1] == str2[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1] + 1;
            } else {
                dp[i][j] = std::max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
    }

    return dp[len_str1][len_str2];
}

File: zig/words/src/words.zig
const std = @import("std");

pub fn reverse_string(str: []const u8) []const u8 {
    const len: usize = str.len;
    var buffer = std.heap.page_allocator.alloc(u8, len) catch {
        return &.{};
    };
    for (str, 0..) |byte, index| {
        buffer[len - index - 1] = byte;
    }
    return buffer;
}

fn removeWhitespaces(str: []const u8, allocator: *const std.mem.Allocator) ![]u8 {
    var buffer = try allocator.alloc(u8, str.len);

    var index: usize = 0;
    for (str) |char| {
        if (!std.ascii.isWhitespace(char)) {
            buffer[index] = char;
            index += 1;
        }
    }
    // Shrink the allocated memory to the exact size of the content.
    // This ensures that what we free later has the same size as what we allocated.
    return allocator.realloc(buffer, index);
}

pub fn is_palindrome(str: []const u8) !bool {
    if (str.len <= 1) {
        return true;
    }
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = &gpa.allocator();
    const cleannedStr = try removeWhitespaces(str, allocator);
    defer allocator.free(cleannedStr);

    var i: usize = 0;
    var j: usize = cleannedStr.len - 1;
    while (i < j) {
        if (cleannedStr[i] != cleannedStr[j]) {
            return false;
        }
        i += 1;
        j -= 1;
    }
    return true;
}

fn computeLpsArray(pattern: []const u8, allocator: *const std.mem.Allocator) ![]usize {
    const len_pattern: usize = pattern.len;
    var lps = try allocator.alloc(usize, len_pattern);
    lps[0] = 0;
    var i: usize = 1;
    var len: usize = 0;
    while (i < len_pattern) {
        if (pattern[i] == pattern[len]) {
            len += 1;
            lps[i] = len;
            i += 1;
        } else {
            if (len > 0) {
                len = lps[len - 1];
            } else {
                lps[i] = 0;
                i += 1;
            }
        }
    }
    return lps;
}

pub fn kmp_search(sub: []const u8, str: []const u8) !bool {
    const len_sub: usize = sub.len;
    const len_str: usize = str.len;
    if (len_sub == 0) {
        return true;
    } else if (len_sub > len_str) {
        return false;
    }
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = &gpa.allocator();
    const lps = try computeLpsArray(sub, allocator);
    defer allocator.free(lps);
    var i: usize = 0;
    var j: usize = 0;
    while (i < len_str) {
        if (str[i] == sub[j]) {
            i += 1;
            j += 1;
            if (j == len_sub) {
                return true;
            }
        } else {
            if (j > 0) {
                j = lps[j - 1];
            } else {
                i += 1;
            }
        }
    }
    return false;
}

pub fn lcs(str1: []const u8, str2: []const u8) !usize {
    const len1: usize = str1.len;
    const len2: usize = str2.len;
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = &gpa.allocator();

    var dp = try allocator.alloc(usize, (len1 + 1) * (len2 + 1));
    defer allocator.free(dp);

    for (0..(len1 + 1)) |i| {
        for (0..(len2 + 1)) |j| {
            dp[i * (len2 + 1) + j] = 0;
        }
    }

    for (1..(len1 + 1)) |i| {
        for (1..(len2 + 1)) |j| {
            if (str1[i - 1] == str2[j - 1]) {
                dp[i * (len2 + 1) + j] = dp[(i - 1) * (len2 + 1) + j - 1] + 1;
            } else {
                dp[i * (len2 + 1) + j] = @max(dp[(i - 1) * (len2 + 1) + j], dp[i * (len2 + 1) + j - 1]);
            }
        }
    }
    return dp[len1 * (len2 + 1) + len2];
}
